### PR TITLE
Duplicates

### PR DESCRIPTION
**Pull request by david-ziegler**
First attempt to detect duplicates (#20)

### CODE REVIEW

File: src/search.rs [hunk 1]
```diff
@@ -51,3 +58,194 @@ fn category_filter_factory<'a>(e: &'a Category) -> Box<Fn(&str) -> bool + 'a> {
...
+
+
+
+
+
+// FIND DUPLICATES:
+
+
+// return vector of entries like: (entry1, entry2, reason) where entry1 and entry2 are similar entries
+pub fn find_duplicates(entries: &Vec<Entry>) -> Vec<(&Entry, &Entry, &str)> {
```
Comment (flosse) [inline src/search.rs]: I'd return just the IDs and the DuplicateType

File: src/search.rs [hunk 2]
```diff
@@ -51,3 +58,194 @@ fn category_filter_factory<'a>(e: &'a Category) -> Box<Fn(&str) -> bool + 'a> {
...
+  //println!("\ncomparing '{}' and '{}':", e1.title, e2.title);
+
+  if similar_title(&e1, &e2, 0.1, 0) && in_close_proximity(&e1, &e2, 100) {Some(DuplicateType::SimilarChars)} 
+  else if similar_title(&e1, &e2, 0.0, 1) && in_close_proximity(&e1, &e2, 100) {Some(DuplicateType::SimilarWords)}
+  else {None}  // entries are not similar 
+}
+
+
+
+fn in_close_proximity(e1: &Entry, e2: &Entry, maxDistMeters:u32) -> bool{
```
Comment (flosse) [inline src/search.rs]: use `max_dist_meters` instead of `maxDistMeters`

File: src/search.rs [hunk 3]
```diff
@@ -51,3 +58,194 @@ fn category_filter_factory<'a>(e: &'a Category) -> Box<Fn(&str) -> bool + 'a> {
...
+    let dist = entry_distance_in_meters(&e1, &e2) as f32;
+    if dist <= maxDistMeters as f32{
+      println!("lat1: {}, lat2: {}, lng1: {}, lng2: {} --> distance: {} m (near)", e1.lat, e2.lat, e1.lng, e2.lng, dist);
```
Comment (flosse) [inline src/search.rs]: use `debug!` or just remove it

File: src/search.rs [hunk 4]
```diff
@@ -51,3 +58,194 @@ fn category_filter_factory<'a>(e: &'a Category) -> Box<Fn(&str) -> bool + 'a> {
...
+fn entry_distance_in_meters(e1: &Entry, e2: &Entry) -> f64{
+  let coord1 = Coordinate{lat: e1.lat, lng:e1.lng};
+  let coord2 = Coordinate{lat: e2.lat, lng:e2.lng};
+  geo::distance(&coord1, &coord2) * 1000.0
+}
+
+
+fn similar_title(e1: &Entry, e2: &Entry, maxPercentDifferent: f32, maxWordsDifferent: u32) -> bool{
+  let maxDist : usize = ((min(e1.title.len(),e2.title.len()) as f32 * maxPercentDifferent) + 1.0) as usize;  // +1 is to get the ceil
+  if hamming_distance_small(&e1.title, &e2.title, maxDist) | (words_equal_except_k_words(&e1.title, &e2.title, maxWordsDifferent)) {
```
Comment (flosse) [inline src/search.rs]: you don't need the `if`...`else` construct :)
..and use `||` instead of `|`

File: src/search.rs [hunk 5]
```diff
@@ -51,3 +58,194 @@ fn category_filter_factory<'a>(e: &'a Category) -> Box<Fn(&str) -> bool + 'a> {
...
+  for i in 0..str1.chars().count() {
+    if str1.chars().nth(i) != str2.chars().nth(i) {
+      dist = dist + 1;
+      if dist > maxDist {
+        break;
+      }
+    }
+  }
+
+  if dist > maxDist{
```
Comment (flosse) [inline src/search.rs]: you don't need the `if` .. `else` construct :)

File: src/main.rs [hunk 6]
```diff
@@ -175,6 +175,35 @@ fn main() {
...
+
+
+    get "/duplicates/" => |req, mut res|{
+      let data: &Data = res.server_data();
+      match data.db_pool()
+        .and_then(|ref pool|{
+
+          Entry::all(pool).map_err(AppError::Store)
+            .and_then(|entries|{
+              let duplicates : Vec<(&Entry, &Entry, &str)> = search::find_duplicates(&entries);  // vec!();
```
Comment (flosse) [inline src/main.rs]: I think you can drop the type definition

File: src/search.rs [hunk 7]
```diff
@@ -51,3 +58,194 @@ fn category_filter_factory<'a>(e: &'a Category) -> Box<Fn(&str) -> bool + 'a> {
...
+
+#[derive(Debug)]
+enum DuplicateType {
+    SimilarChars,
+    SimilarWords
+}
+// returns 0 if e1 and e2 are not similar_titl
+// returns 1 if the hamming distance of their titles is small
+// returns 2 if at most one word of their titles is different (and they don't both have one-word titles)
+// returns 3 if they are located within 10 meters of each other
```
Comment (flosse) [inline src/search.rs]: I think these comments are out of date ;-)

File: src/search.rs [hunk 8]
```diff
@@ -1,6 +1,13 @@
 // Copyright (c) 2015 Markus Kohlhase <mail@markus-kohlhase.de>
 
+use geo;
+
 use json::{Entry, Category};
+use geo::Coordinate;
+use std::cmp::min;
+use std::collections::HashSet;
+use std::f64::consts;
```
Comment (flosse) [inline src/search.rs]: `consts` can be dropped

File: src/main.rs [hunk 9]
```diff
@@ -175,6 +175,35 @@ fn main() {
         }
       }
 
+
+
```
Comment (flosse) [inline src/main.rs]: please remove empty lines

File: src/search.rs [hunk 10]
```diff
@@ -51,3 +57,275 @@ fn category_filter_factory<'a>(e: &'a Category) -> Box<Fn(&str) -> bool + 'a> {
...
+  let coord1 = Coordinate{lat: e1.lat, lng:e1.lng};
+  let coord2 = Coordinate{lat: e2.lat, lng:e2.lng};
+  geo::distance(&coord1, &coord2) * 1000.0
+}
+
+
+fn similar_title(e1: &Entry, e2: &Entry, maxPercentDifferent: f32, maxWordsDifferent: u32) -> bool{
+  let maxDist : usize = ((min(e1.title.len(),e2.title.len()) as f32 * maxPercentDifferent) + 1.0) as usize;  // +1 is to get the ceil
+  
+  levenshtein_distance_small(&e1.title, &e2.title, maxDist) | (words_equal_except_k_words(&e1.title, &e2.title, maxWordsDifferent))
```
Comment (flosse) [inline src/search.rs]: please use `||` instead of `|`

**Review by flosse — CHANGES_REQUESTED** [review]
I'm looking forward :)

**Comment by david-ziegler** [comment]
thanks for the comments! I learned something from it :)

I made one other change: I noticed that for levenshtein-distance I earlier copied the naive algorithm from wikipedia. That calculates distances over and over again. Now I implemented the dynamic programming version, which is faster. It's still not as fast as hamming-distance (it takes up to 10 seconds for the whole dataset) but I guess that's more or less acceptable. Levenshtein-distance is better to capture small typos or so. E.g. "a12345" and "12345" have hamming-distance 6 but levenshtein-distance 1.

**Comment by flosse** [comment]
I like the levenshtein distance :) 
And i think there are good chances to speed up the calculation. But even if not, it doesn't matter because this method will be called once a week or month, so we shouldn't care too much ;-)

**Comment by flosse** [comment]
...moreover we could use both: e.g. the web app calls `/entries/duplicates?type=simple` for the first run and presents it to the user. Meanwhile a second call `/entries/duplicates?type=precise` gets fired to improve the results.

**Comment by flosse** [comment]
I think we could merge this soon.
Would you mind to rebase your PR into one commit?

**Review by flosse — APPROVED** [review]

**Pull request merged**